packages/x-license/src/utils/licenseScope.ts

```diff
@@ -1,5 +1,24 @@
-export const LICENSE_SCOPES = ['pro', 'premium'] as const;
+export const LICENSE_SCOPES = ['pro', 'premium', 'pro2024', 'premium2024'] as const;
```

Comment:
We must be careful with backward compatibility here.

Modifying the existing list of scopes is risky because it will prevent people with new licenses to use "old" versions of the software unless they are grandfathered.

The vast majority of people are not impacted by the cap, so they can just upgrade to the new license format without any cost change. Which means, grand-fathering will only impact a fraction of our sales.
But if you put the information in the scope, it means that anyone that do not use a version of the package published after the release of this PR needs to use a grandfathered license, which increases a lot the amount of grandfathered licenses since most people are not using the last version.
This also means that if they want to use the charts of the tree view in the future, the will have to ask for a new license key which adds a lot of work on our customer success team and an extra step for the users.

My proposal is to add this information in a new property to keep the new licenses parsable by older version of the software.

```tsx
export const LICENSE_SCOPES = ['pro', 'premium'] as const;

// could be
export const LICENSE_PRICING_VERSION = [1, 2] as const;
// or
export const LICENSE_PRICING_VERSION = ['initial', 'Q3-2024'] as const;
// or anything was of representing it that feels best for us
```

The data present in the license is the same (you use `2024`, Jose proposed to add the quarter in case we have 2 pricing changes in one year, but it's a detail), but it's not placed in the same properties in order to guarantee backward compatibility.
And for future pricing changes, we should be able to always use this new property. Imagine we release a new version of the pricing that adds back a cap at 50 seats, we can add some rules based on this property.

____


From a product perspective, I think the difference is whether we consider these new pricing to be new products or evolution of the same products.
For me it's an evolution of the same product since we want everybody to eventually migrate to it, we have to grandfather, but we are putting limits to encourage existing customers to migrate to it.
IMHO the goal is even to drop support for the old approach in some future major (people could renew their grandfathered license and keep the cap, but they would not be able to upgrade to a new version of the package). Otherwise, as the pricing evolves, our licensing model will become impossible to understand with tons of different offers depending on when the users bought its first license.

Reply:
but in this case we would alter the license structure:

```diff
-Format: O=${orderNumber},E=${expiryTimestamp},S=${scope},LM=${licensingModel},KV=2`;
+Format: O=${orderNumber},E=${expiryTimestamp},S=${scope},LM=${licensingModel},PV=${pricingVersion},KV=2`;
```

otherwise we would have no way of telling if we are dealing with an updated pricing plan or not when decoding a license.

---

That being said your comments are valid and I do understand that, but wouldn't that essentially mean we are heading towards a new keyversion then?

Reply:
Or we do make this additional property optional:

```diff
-Format: O=${orderNumber},E=${expiryTimestamp},S=${scope},LM=${licensingModel},KV=2`;
+Format: O=${orderNumber},E=${expiryTimestamp},S=${scope},LM=${licensingModel}[,PV=${pricingVersion}],KV=2`;
```

And then in decoding:

```tsx
if (key === 'PV') {
  licenseInfo.pricingVersion = value;
}
```

Then take it from there without storing it in the global licensing object in `window` 🤔

Reply:
If you add the `PV` to the license generation and validation.
Then all licenses generated after your PR is released have `PV`, older do not have it
And we have the following behavior:

- If a license with `PV` is parsed with a newer version of the package, it uses `PV` to determined if the charts and tree view can be accessed
- If a license with `PV` is parsed with an older version of the package, it does not know about PV so it will just ignore it. `KV=2` was built so that we can add properties. Ifyou look at `decodeLicenseVersion2`, we loop through all the properties and just ignore the one we don't know. So every package will be accessible, but since charts and tree view are not released yet it's not a problem.
- If a license without `PV` is parsed with a newer version of the package, it puts a default value to it (equal to however we name the old version of the model, `2023`, `initial`, `1`, etc...) so it won't be able to use charts and tree view
- If a license without `PV` is parsed with an older version of the package, it will act just like today and give access to all packages, but since charts and tree view are not released yet it's not a problem.


For me all scenarios works just fine.

Reply:
At the parsing level, for me the only change we need is:

```tsx
@@ -73,6 +74,10 @@ const decodeLicenseVersion2 = (license: string): MuiLicense => {
           licenseInfo.expiryTimestamp = expiryTimestamp;
         }
       }
+
+      if (key === 'PV') {
+        licenseInfo.planVersion = value;
+      }
     });
 
   return licenseInfo;
```

Reply:
At the generation level: 

```diff
diff --git a/packages/x-license/src/generateLicense/generateLicense.ts b/packages/x-license/src/generateLicense/generateLicense.ts
index 9ff33495f..eb747b73f 100644
--- a/packages/x-license/src/generateLicense/generateLicense.ts
+++ b/packages/x-license/src/generateLicense/generateLicense.ts
@@ -10,6 +10,7 @@ export interface LicenseDetails {
   expiryDate: Date;
   scope: LicenseScope;
   licensingModel: LicensingModel;
+  planVersion: PlanVersion;
 }
 
 function getClearLicenseString(details: LicenseDetails) {
@@ -23,7 +24,7 @@ function getClearLicenseString(details: LicenseDetails) {
 
   return `O=${details.orderNumber},E=${details.expiryDate.getTime()},S=${details.scope},LM=${
     details.licensingModel
-  },KV=${licenseVersion}`;
+  },PV=${details.planVersion},KV=${licenseVersion}`;
 }
 
 export function generateLicense(details: LicenseDetails) {
```

Reply:
And then the bulk of the changes is at the `verifyLicense` level.
I did not check your changes their :+1:

Reply:
> That being said your comments are valid and I do understand that, but wouldn't that essentially mean we are heading towards a new keyversion then?

The `KV` was only introduced because the 1st version of the key parsing was based on a RegExp and any change on the license was basically non backward compatible. When we created the premium plan, we had to add the `scope`, and we took the opportunity to rename the keys to initials in order to shorten the license. But this meant changing the algorithm that parses the license.

`KV=2` just loops on all the properties as long as they have the format `{key}={value},{key}={value}......`.
We might have to do some breaking changes on the licenses at some point, but tbh I think this structure allows us to do quite a lot of changes in a non breaking way.

Reply:
> At the parsing level, for me the only change we need is:
> 
> ```diff
> diff --git a/packages/x-license/src/verifyLicense/verifyLicense.ts b/packages/x-license/src/verifyLicense/verifyLicense.ts
> index b280b196d..4a87eec24 100644
> --- a/packages/x-license/src/verifyLicense/verifyLicense.ts
> +++ b/packages/x-license/src/verifyLicense/verifyLicense.ts
> @@ -52,6 +52,7 @@ const decodeLicenseVersion2 = (license: string): MuiLicense => {
>      scope: null,
>      licensingModel: null,
>      expiryTimestamp: null,
> +    planVersion: '1'
>    };
>  
>    license
> @@ -73,6 +74,10 @@ const decodeLicenseVersion2 = (license: string): MuiLicense => {
>            licenseInfo.expiryTimestamp = expiryTimestamp;
>          }
>        }
> +      
> +      if (key === 'PV') {
> +        licenseInfo.planVersion = value;
> +      }
>      });
>  
>    return licenseInfo;
> ```

Yes, that is what I came up with as well.

Reply:
> At the generation level:
> 
> ```diff
> diff --git a/packages/x-license/src/generateLicense/generateLicense.ts b/packages/x-license/src/generateLicense/generateLicense.ts
> index 9ff33495f..eb747b73f 100644
> --- a/packages/x-license/src/generateLicense/generateLicense.ts
> +++ b/packages/x-license/src/generateLicense/generateLicense.ts
> @@ -10,6 +10,7 @@ export interface LicenseDetails {
>    expiryDate: Date;
>    scope: LicenseScope;
>    licensingModel: LicensingModel;
> +  planVersion: PlanVersion;
>  }
>  
>  function getClearLicenseString(details: LicenseDetails) {
> @@ -23,7 +24,7 @@ function getClearLicenseString(details: LicenseDetails) {
>  
>    return `O=${details.orderNumber},E=${details.expiryDate.getTime()},S=${details.scope},LM=${
>      details.licensingModel
> -  },KV=${licenseVersion}`;
> +  },PV=${details.planVersion},KV=${licenseVersion}`;
>  }
>  
>  export function generateLicense(details: LicenseDetails) {
> ```

same as this one.
only difference to increase readabilty:

```tsx
const keyParts = [
  `O=${details.orderNumber}`,
  `E=${details.expiryDate.getTime()}`,
  `S=${details.scope}`,
  `LM=${details.licensingModel}`,
  `KV=${licenseVersion}`,
];

if (details.planVersion) {
  keyParts.splice(3, 0, `PV=${details.planVersion}`);
}

return keyParts.join(',');
```

Reply:
I don't understand what you mean by optional
Can't we just always add the `planVersion` on newly generated licenses?

Reply:
i meant optional when decoding. Of course we will always add it when creating a new license! ... sry if that wasn't clear

Reply:
OK, I saw your modified code snippet, I do agree that if would be more readable with a `join` :+1: